v2/component/launcher.go

```diff
@@ -0,0 +1,440 @@
+	PipelineRunID     string
+	PipelineTaskID    string
+	TaskName          string
+	ContainerImage    string
+	MLMDServerAddress string
+	MLMDServerPort    string
+}
+
+type bucketConfig struct {
+	scheme     string
+	bucketName string
+	prefix     string
+}
+
+func (b *bucketConfig) bucketURL() string {
+	u := b.scheme + b.bucketName
+
+	if len(b.prefix) > 0 {
+		u = fmt.Sprintf("%s?prefix=%s", u, b.prefix)
+	}
+	return u
+}
+
+func (b *bucketConfig) keyFromURI(uri string) (string, error) {
+	prefixedBucket := b.scheme + path.Join(b.bucketName, b.prefix)
+	if !strings.HasPrefix(uri, prefixedBucket) {
+		return "", fmt.Errorf("URI %q does not have expected bucket prefix %q", uri, prefixedBucket)
+	}
+
+	key := strings.TrimLeft(strings.TrimPrefix(uri, prefixedBucket), "/")
+	if len(key) == 0 {
+		return "", fmt.Errorf("URI %q has empty key given prefixed bucket %q", uri, prefixedBucket)
+	}
+	return key, nil
+}
+
+func (b *bucketConfig) uriFromKey(blobKey string) string {
+	return b.scheme + path.Join(b.bucketName, b.prefix, blobKey)
+}
+
+var bucketPattern = regexp.MustCompile(`^([a-z][a-z0-9]+:///?)([^/]+)(/[^ ?]*)?$`)
+
+func parseBucketConfig(path string) (*bucketConfig, error) {
+	ms := bucketPattern.FindStringSubmatch(path)
+	if ms == nil || len(ms) != 4 {
+		return nil, fmt.Errorf("Unrecognized pipeline root format: %q", path)
+	}
+
+	// TODO: Verify/add support for s3:// and file:///.
+	if ms[1] != "gs://" {
+		return nil, fmt.Errorf("Unsupported Cloud bucket: %q", path)
+	}
+
+	prefix := strings.TrimPrefix(ms[3], "/")
+	if len(prefix) > 0 && !strings.HasSuffix(prefix, "/") {
+		prefix = prefix + "/"
+	}
+
+	return &bucketConfig{
+		scheme:     ms[1],
+		bucketName: ms[2],
+		prefix:     prefix,
+	}, nil
+}
+
+func (o *LauncherOptions) validate() error {
+	empty := func(s string) bool { return len(s) == 0 }
+	err := func(s string) error { return fmt.Errorf("Must specify %s", s) }
+
+	if empty(o.PipelineName) {
+		return err("PipelineName")
+	}
+	if empty(o.PipelineRunID) {
+		return err("PipelineRunID")
+	}
+	if empty(o.PipelineTaskID) {
+		return err("PipelineTaskID")
+	}
+	if empty(o.PipelineRoot) {
+		return err("PipelineRoot")
+	}
+	if empty(o.TaskName) {
+		return err("TaskName")
+	}
+	if empty(o.MLMDServerAddress) {
+		return err("MLMDServerAddress")
+	}
+	if empty(o.MLMDServerPort) {
+		return err("MLMDServerPort")
+	}
+	return nil
+}
+
+// NewLauncher creates a new launcher object using the JSON-encoded runtimeInfo
+// and specified options.
+func NewLauncher(runtimeInfo string, options *LauncherOptions) (*Launcher, error) {
+	if err := options.validate(); err != nil {
+		return nil, err
+	}
+
+	bc, err := parseBucketConfig(options.PipelineRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	rt, err := parseRuntimeInfo(runtimeInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	// Placeholder replacements.
+	pr := make(map[string]string)
+
+	metadataClient, err := metadata.NewClient(options.MLMDServerAddress, options.MLMDServerPort)
```

Comment:
nit: probably do not matter much now, but is the initiated grpc connection going to take up resources in MLMD?

Would it be better to do the following instead?

1. start MLMD connection #1
2. bookkeep execution and artifacts
3. stop connection #1
4. run user command
5. start connection #2
6. ...

Reply:
I think that's an optimization we can make if needed. I don't think these take much resources though. And for quick jobs, it's cheaper to keep the channel around.

Reply:
I thought the gRPC server has a limit on concurrent connections, if that's not the case, totally fine with it.